src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/Intermediate/DocumentIntermediateNodeExtensions.cs

```diff
@@ -42,94 +44,111 @@ public static NamespaceDeclarationIntermediateNode FindPrimaryNamespace(this Doc
 
     public static IReadOnlyList<IntermediateNodeReference> FindDirectiveReferences(this DocumentIntermediateNode node, DirectiveDescriptor directive)
     {
-        if (node == null)
-        {
-            throw new ArgumentNullException(nameof(node));
-        }
+        ArgHelper.ThrowIfNull(node);
+        ArgHelper.ThrowIfNull(directive);
 
-        if (directive == null)
-        {
-            throw new ArgumentNullException(nameof(directive));
-        }
+        using var results = new PooledArrayBuilder<IntermediateNodeReference>();
+        node.CollectDirectiveReferences(directive, ref results.AsRef());
 
-        var visitor = new DirectiveVisitor(directive);
-        visitor.Visit(node);
-        return visitor.Directives;
+        return results.ToImmutableAndClear();
     }
 
-    public static IReadOnlyList<IntermediateNodeReference> FindDescendantReferences<TNode>(this DocumentIntermediateNode document)
-        where TNode : IntermediateNode
+    internal static void CollectDirectiveReferences(this DocumentIntermediateNode document, DirectiveDescriptor directive, ref PooledArrayBuilder<IntermediateNodeReference> references)
```

Comment:
>DirectiveDescriptor [](http://example.com/codeflow?start=92&length=19)

Dumb question: There's no side information stored on the node or a cwt that indicates whether a directive is contained within, right? I know roslyn has some information stored on the green nodes like whether directives/attributes/diagnostics are contained within, but I don't know if razor has a similar concept on the node.

Reply:
Green nodes in Razor have a flag for diagnostics, but not for directives. However, these are the intermediate representation nodes, which are much less evolved than the green/red syntax node model. So, the answer to your question is "yes, Razor does have such flags" but also, "not here." :smile: